src/aics-image-viewer/components/TfEditor/index.tsx

```diff
@@ -143,68 +179,93 @@ const TfEditor: React.FC<TfEditorProps> = (props) => {
     }
+  };
 
-    setSelectedPointIdx(draggedPointIdxRef.current);
+  const handlePlotPointerDown: React.PointerEventHandler<SVGSVGElement> = (event) => {
+    if (props.useControlPoints) {
+      // Advanced mode - we're either creating a new control point or selecting/dragging an existing one
+      if (draggedPointIdxRef.current === null && event.button === 0) {
+        // this click is not on an existing point - create a new one
+        const [x, opacity] = mouseEventToControlPointValues(event);
+        const point = { x, opacity, color: lastColorRef.current };
+
+        // add new control point to controlPoints
+        const index = d3.bisector<ControlPoint, ControlPoint>((a, b) => a.x - b.x).left(props.controlPoints, point);
+        setDraggedPointIdx(index);
+
+        const newControlPoints = [...props.controlPoints];
+        newControlPoints.splice(index, 0, point);
+        setControlPoints(newControlPoints);
+      }
 
-    if (event.button === 0) {
+      if (typeof draggedPointIdxRef.current !== "string") {
+        setSelectedPointIdx(draggedPointIdxRef.current);
+      }
+    }
+
+    if (event.button === 0 && draggedPointIdxRef.current !== null) {
       // get set up to drag the point around, even if the mouse leaves the SVG element
       event.currentTarget.setPointerCapture(event.nativeEvent.pointerId);
     } else {
       setDraggedPointIdx(null);
     }
   };
 
-  const handleControlPointDrag: React.PointerEventHandler<SVGSVGElement> = (event) => {
+  const handlePlotPointerMove: React.PointerEventHandler<SVGSVGElement> = (event) => {
     if (draggedPointIdxRef.current === null) {
       return;
     }
+
     if ((event.buttons & MOUSE_EVENT_BUTTONS_PRIMARY) === 0) {
-      handleControlPointDragEnd(event);
+      handleDragEnd(event);
       return;
     }
-    const draggedIdx = draggedPointIdxRef.current;
+
     event.stopPropagation();
     event.preventDefault();
-
-    // Update dragged control point
     const [x, opacity] = mouseEventToControlPointValues(event);
-    const newControlPoints = [...controlPointsRef.current];
-    const draggedPoint = newControlPoints[draggedIdx];
-    draggedPoint.x = x;
-    draggedPoint.opacity = opacity;
-
-    // Remove control points to keep the list sorted by x value
-    const bisector = d3.bisector<ControlPoint, ControlPoint>((a, b) => a.x - b.x);
-    const idxLeft = bisector.left(newControlPoints, draggedPoint);
-    const idxRight = bisector.right(newControlPoints, draggedPoint);
 
-    if (idxLeft < draggedIdx) {
-      const numPointsToRemove = draggedIdx - idxLeft; // should almost always be 1
-      newControlPoints.splice(idxLeft, numPointsToRemove);
-      draggedPointIdxRef.current -= numPointsToRemove;
-      setSelectedPointIdx(draggedPointIdxRef.current);
-    } else if (idxRight > draggedIdx + 1) {
-      newControlPoints.splice(draggedIdx + 1, idxRight - draggedIdx - 1);
+    if (typeof draggedPointIdxRef.current === "number") {
```

Comment:
what is the other type that's not a "number"?  should you just check props.useControlPoints here instead?  or maybe `draggedPointIdxRef.current as number` vs `draggedPointIdxRef.current as minmaxsliderenumthingy` in the lines below - for readability

Reply:
I agree that `props.useControlPoints` would make more semantic sense here. But it also wouldn't cause TypeScript to narrow this type for the following lines, where we either pass `draggedPointIdxRef.current` to a function that takes a `number` or one that takes a `TfEditorRampSliderHandle`.